openstack/messaging/v2/queues/requests.go

```diff
@@ -0,0 +1,158 @@
+	return pager
+}
+
+// CreateOptsBuilder allows extensions to add additional parameters to the
+// Create request.
+type CreateOptsBuilder interface {
+	ToQueueCreateMap() (map[string]interface{}, error)
+}
+
+// CreateOpts specifies the queue creation parameters.
+type CreateOpts struct {
+	// The name of the queue to create.
+	QueueName string `json:"queue_name" required:"true"`
+
+	// The target incoming messages will be moved to when a message can’t
+	// processed successfully after meet the max claim count is met.
+	DeadLetterQueue string `json:"_dead_letter_queue,omitempty"`
+
+	// The new TTL setting for messages when moved to dead letter queue.
+	DeadLetterQueueMessagesTTL int `json:"_dead_letter_queue_messages_ttl,omitempty"`
+
+	// The delay of messages defined for a queue. When the messages send to
+	// the queue, it will be delayed for some times and means it can not be
+	// claimed until the delay expired.
+	DefaultMessageDelay int `json:"_default_message_delay,omitempty"`
+
+	// The default TTL of messages defined for a queue, which will effect for
+	// any messages posted to the queue.
+	DefaultMessageTTL int `json:"_default_message_ttl" required:"true"`
+
+	// The flavor name which can tell Zaqar which storage pool will be used
+	// to create the queue.
+	Flavor string `json:"_flavor,omitempty"`
+
+	// The max number the message can be claimed.
+	MaxClaimCount int `json:"_max_claim_count,omitempty"`
+
+	// The max post size of messages defined for a queue, which will effect
+	// for any messages posted to the queue.
+	MaxMessagesPostSize int `json:"_max_messages_post_size,omitempty"`
+
+	// Extra is free-form extra key/value pairs to describe the queue.
+	Extra map[string]interface{} `json:"-"`
+}
+
+// ToQueueCreateMap constructs a request body from CreateOpts.
+func (opts CreateOpts) ToQueueCreateMap() (map[string]interface{}, error) {
+	b, err := gophercloud.BuildRequestBody(opts, "")
+	if err != nil {
+		return nil, err
+	}
+
+	if opts.Extra != nil {
+		for key, value := range opts.Extra {
+			b[key] = value
+		}
+
+	}
+	return b, nil
+}
+
+// Create requests the creation of a new queue.
+func Create(client *gophercloud.ServiceClient, clientID string, opts CreateOptsBuilder) (r CreateResult) {
+	b, err := opts.ToQueueCreateMap()
+	if err != nil {
+		r.Err = err
+		return
+	}
+
+	queueName := b["queue_name"].(string)
+	delete(b, "queue_name")
+
+	_, r.Err = client.Put(createURL(client, queueName), b, r.Body, &gophercloud.RequestOpts{
+		OkCodes:     []int{201, 204},
+		MoreHeaders: map[string]string{"Client-ID": clientID},
+	})
+	return
+}
+
+// UpdateOptsBuilder allows extensions to add additional parameters to the
+// update request.
+type UpdateOptsBuilder interface {
+	ToQueueUpdateMap() (map[string]interface{}, error)
+}
+
+// UpdateOpts is an array of UpdateQueueBody.
+type UpdateOpts []UpdateQueueBody
+
+// UpdateOpts implements UpdateOpts.
+type UpdateQueueBody struct {
+	Op    string      `json:"op" required:"true"`
```

Comment:
There are probably a finite amount of ops, right? I did a quick search and I think they are "add", "remove", and "replace":

https://github.com/openstack/zaqar/blob/83300343d23051db31702d1cadbf9e250c37ef24/zaqar/transport/wsgi/v2_0/queues.py#L221-L244

It would be best to have a custom type of UpdateOp and then have each type as a const. Something like:

```go
type UpdateOp string
const (
  ReplaceOp UpdateOp = "replace"
  AddOp UpdateOp = "add"
  RemoveOp UpdateOp = "remove"
)
```